### ISSUE TITLE
Cannot use NumPy ints for register indices

### ISSUE THREAD

**Issue by nonhermitian** [Qiskit/qiskit-terra#2291]
### Information

- **Qiskit Terra version**: master
- **Python version**:
- **Operating system**: osx

### What is the current behavior?
```python

import numpy as np
from qiskit import *

arr = np.array([0])
q = QuantumRegister(1,'q')
qc = QuantumCircuit(q)
qc.h(q[arr[0]])
```

gives

```python
     87         """
     88         if not isinstance(key, (int, slice, list)):
---> 89             raise QiskitError("expected integer or slice index into register")
     90         if isinstance(key, int) and key < 0:
     91             key = self.size + key

QiskitError: 'expected integer or slice index into register'
```

This is similar to #2152 

### What is the expected behavior?
I would expect this to work.  Especially since functions like `np.random.randint()` and `np.random.permutation()` come in handy when building random circuits, e.g. Quantum Volume.

I label this as a bug because we should be compatible with the dtypes from the foundational numerical Python library.

### Suggested solutions
Similar to the one in #2152 by @chriseclectic

**Comment by 1ucian0** [comment]
I think this casting to python ints should be on the user side. Numpy is too out of our control to make us dependant on it.

Our methods are already very polymorphic. With each new input type, they are getting harder and harder to maintain.

**Comment by nonhermitian** [comment]
I believe we are already dependent on numpy. In the scientific python community, everyone plays nice with numpy. This is simply extending the isinstance check(s)

**Comment by 1ucian0** [comment]
We check for `int` type 54 times in code.

**Comment by nonhermitian** [comment]
Ok cool. Now we know how many things need to be modified.

**Comment by 1ucian0** [comment]
Parameters can also be lists... that means that we should also support ` circuit.h(np.array([0,1]))`? What about `circuit.h(np.array([[0,1],[2,3]]))`? The answer can be yes to everything, but there is a performance and maintainability price to pay here...

**Comment by kdk** [comment]
I don't see why we can't follow an approach similar to #2288 here, especially after the consolidation in #2282 . I agree that users having a np.int is probably common enough, and its clear what the expected behavior would be.

**Comment by 1ucian0** [comment]
It would not be as clean as in #2288 because the extremely polymorphic nature of gate arguments. But if the line is in numpy ints, we can explore options. I'm open to be convinced. Shall the issue be reopen or does @nonhermitian decided was a bad idea?

**Comment by 1ucian0** [comment]
I'm just worried to open the numpy-type-hell door to an already very cluttered API.

**Comment by nonhermitian** [comment]
Your comment is not very constructive. The NumPy array typing is very logical, and extensible, allowing you to leverage fast memory-aligned C-arrays via Python.  It is incredibly fast, and allows for array indexing tricks like slicing that make no memory copies.  These benefits, among others, is why every single scientific package in Python (including this one) uses it.  I have closed this issue because you do not seem to want users who use the tools in NumPy to have a seamless experience interfacing with your own software.

**Comment by kdk** [comment]
@1ucian0 's concern is not a baseless one (considering the difficulty of #2282) and adding support for numpy types in a maintainable way would require some thinking, but I certainly don't think its insurmountable, and given the community interest (e.g #2152, #1898, #2056, #2059, #2151, ...) it seems worth pursuing.

**Comment by 1ucian0** [comment]
I fully agree that terra should "plays nice with numpy". As I said in https://github.com/Qiskit/qiskit-terra/issues/2152#issuecomment-489111940, probably the cast-before-validate approach can be a good natural limit in the API's input types.

However, I would like to emphasis that a limit in the input type is inevitable (and repeat that there is a performance and maintenance price for casting and type checks). 

I will try to add the idea of a best-effort casting in #2282. Reopening while there is a discussion.

**Comment by 1ucian0** [comment]
The best-effort-type-casting approach:
https://github.com/1ucian0/qiskit-terra/commit/6b36e5dedac4b62f33db34ae1e2a47d189a1ed11

**Comment by 1ucian0** [comment]
closed via #2288
